046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,64 @@
+# Analyse broker metrics before rolling Kafka
+
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery.
+
+## Current situation
+
+The current logic in KafkaRoller:
+- Take a list of Kafka pods and reorder them based on the readiness.
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a Kafka pod, consider if it is a controller or if it has an impact on the availability.
+- If a Kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then force restart it without considering why it's taking a long time.
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+ 
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should wait longer instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in and avoids unnecessary restarts.
+ 
+1. Add a RESTful endpoint to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on port 8443.
+    - It will reuse the broker's own certificate, which will already be in the container.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of the `BrokerState` metric.
```

Comment:
I suggested using the Servlet API mainly because it was already used in another Strimzi operator ( https://github.com/strimzi/kafka-access-operator/blob/main/src/main/java/io/strimzi/kafka/access/server/HealthServlet.java). However, happy to change it to  use Jetty's native ServerHandler API.